### PR TITLE
Model training 

### PR DESCRIPTION
**Pull request by rasswanth-s**
## Description
Please include a summary of the change, the motivation, and any additional context that will help others understand your PR. If it closes one or more open issues, [please tag them as described here](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

## Affected Dependencies
List any dependencies that are required for this change.

## How has this been tested?
- Describe the tests that you ran to verify your changes.
- Provide instructions so we can reproduce.
- List any relevant details for your test configuration.

## Checklist
- [ ] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [ ] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ ] My changes are covered by tests

### CODE REVIEW

**Comment by iamtrask** [comment]
This PR is much too large to be merged. Let's split it into sections, starting with a PR importing the tensor type improvements. 

We need to have more discussion about the deep learning library related objects which, if pulled in, could trigger a lot of longer term support which may or may not be in our best interests. I do, however, think we've prototyped here some techniques which could apply to the PyTorch layer types once they're ready.

**Review by iamtrask — CHANGES_REQUESTED** [review]
This PR is a bit on the large side to be merged as is (might take a long time to get everything reviewed all at once). Let's split it into sections, starting with a PR importing the tensor type improvements (ideally integrating with previous versions of these tensors instead of replacing them entirely, as they have different performance characteristics).

Also let's have more discussion about the deep learning library related objects. It's clear we've prototyped here some techniques which could apply to the PyTorch layer types once they're ready, so the meta-work here on deep learning library support is likely a valuable start. It's unclear to me what the upgrade path is from our custom layer types here to PyTorch ones. So that'll be a good topic for discussion.

It may be that we decide to merge in these layer types, but articulate that these are temporary until PyTorch ones are ready. If this was the decision, we should put flags/warnings to let people know that the custom->pytorch layer type transition might have some slight API changes coming.

File: packages/syft/src/syft/core/tensor/autodp/gamma_tensor.py [hunk 1]
```diff
@@ -1317,26 +1426,68 @@ def _gt(state: dict) -> jax.numpy.DeviceArray:
...
+            data_subjects=output_ds,
+            min_vals=min_val,
+            max_vals=max_val,
             func=_gt,
             state=output_state,
         )
 
+    def __le__(self, other: Any) -> GammaTensor:
+        # relative
+        from .phi_tensor import PhiTensor
```
Comment (iamtrask) [inline packages/syft/src/syft/core/tensor/autodp/gamma_tensor.py:1453]: @madhavajay - we need to figure out a class hierarchy where we don't have any more of these relative imports. In this case we only need it as an argument type checker. What's the right paradigm here? My guess is we need AbstractPhiTensor rand AbstrractGammaTensor to avoid the circular dependency of each tensor needing to check arguments to see if its the other type.

File: packages/syft/src/syft/core/tensor/autodp/gamma_tensor.py [hunk 2]
```diff
@@ -1317,26 +1426,68 @@ def _gt(state: dict) -> jax.numpy.DeviceArray:
...
+            child = self.child.__le__(other.child)
+            output_ds = self.data_subjects <= other.data_subjects
+
+        else:
+
+            def _le(state: dict) -> jax.numpy.DeviceArray:
+                return jnp.less_equal(self.run(state), other)
+
+            child = self.child.__le__(other)
+            output_ds = self.data_subjects
```
Comment (iamtrask) [inline packages/syft/src/syft/core/tensor/autodp/gamma_tensor.py:1477]: See above. Data subjects should have control here.

File: packages/syft/src/syft/core/tensor/autodp/gamma_tensor.py [hunk 3]
```diff
@@ -1353,23 +1504,59 @@ def exp_reduction(val: np.ndarray) -> np.ndarray:
...
+            )
+
+        def _log(state: dict) -> jax.numpy.DeviceArray:
+            return jnp.log(self.run(state))
+
+        return GammaTensor(
+            child=np.log(self.child),
+            min_vals=min_val,
+            max_vals=max_val,
+            data_subjects=self.data_subjects,
```
Comment (iamtrask) [inline packages/syft/src/syft/core/tensor/autodp/gamma_tensor.py:1555]: See above. Data subjects should have control here.

**Review by iamtrask — CHANGES_REQUESTED** [review]
Lots of good work here. I've done a first pass on a review, mostly looking at the DSL and how it could best stitch in.

File: packages/syft/src/syft/core/node/common/client.py [hunk 4]
```diff
@@ -510,7 +510,9 @@ def get(self, key: Union[str, int, UID]) -> Pointer:
                 address=self.client.address, reply_to=self.client.address, obj_id=key
             )
             results = getattr(
-                self.client.send_immediate_msg_with_reply(msg=msg), "results", None
+                self.client.send_immediate_msg_with_reply(msg=msg, timeout=600),
```
[the remark this answers is not in the document]
Comment (shubham3121) [reply packages/syft/src/syft/core/node/common/client.py:513]: @iamtrask, it's 30 seconds by default. But it may be not be ideal for get or upload data (since size of data can be large and could take time to download). I move these timeouts to constant variables to its easy to read and manipulate those values.

File: packages/syft/src/syft/core/tensor/autodp/gamma_tensor.py [hunk 5]
```diff
@@ -1278,23 +1385,25 @@ def _rmatmul(state: dict) -> jax.numpy.DeviceArray:
...
+            min_val = self.min_vals.__rmatmul__(other)
+            max_val = self.max_vals.__rmatmul__(other)
+            output_ds = self.data_subjects.__rmatmul__(other)
 
         return GammaTensor(
             child=child,
-            data_subjects=self.data_subjects,
-            min_val=min_val,
-            max_val=max_val,
+            data_subjects=output_ds,
```
[the remark this answers is not in the document]
Comment (shubham3121) [reply packages/syft/src/syft/core/tensor/autodp/gamma_tensor.py:1404]: here they do, `output_ds = self.data_subjects.__rmatmul__(other)`

**Comment by iamtrask** [comment]
Impressive pace of progress in this PR. Also great job breaking things into lots of intuitive commits.

**Comment by iamtrask** [comment]
https://github.com/OpenMined/PySyft/pull/6608#pullrequestreview-1033405428 Upon further consideration and disussion - it seems more important that 0.7 be able to fully execute the MFLP demo. So we should lean towards merging in all of the new functionality once its ready (and tests are passing). 0.8 is a significant refactor push during which we can dig into the deeper issues, and it seems like there's a smooth upgrade path from the functionality here (0.7) and functionality we intend to support in the future.

Please forgive my message before which now feels pre-mature and not fully considered. I was not at my best.

File: packages/syft/src/syft/core/adp/data_subject_list.py [hunk 6]
```diff
@@ -115,30 +147,674 @@ def from_objs(entities: Union[np.ndarray, list]) -> DataSubjectList:
...
+        dsl1_target_shape = (
+            (*dsl1.shape[:-1], 1) if len(dsl1.shape) == 2 else (*dsl1.shape[1:-1], 1)
+        )
+        dsl2_target_shape = (
+            (1, *dsl2.shape[:-3], dsl2.shape[-1])
+            if len(dsl2.shape) == 2
+            else (*dsl2.shape[1:-2], 1, dsl2.shape[-1])
+        )
+
+        summed_dsl1 = dsl1.sum(target_shape=dsl1_target_shape)
```
[the remark this answers is not in the document]
Comment (rasswanth-s) [reply packages/syft/src/syft/core/adp/data_subject_list.py]: Currently the DataSubjectList is stale code, 
We have moved to DataSubjectArray , where we treat it as a numpy array, 
and we could call .sum() like we do in a numpy array.

[the remark this answers is not in the document]
Comment (rasswanth-s) [reply packages/syft/src/syft/core/adp/data_subject_list.py]: We plan to remove DataSubjectList in the next PR, so we get it quickly merge with new DataSubjectArray(DSA)

File: packages/syft/src/syft/core/adp/data_subject_list.py [hunk 7]
```diff
@@ -115,30 +147,674 @@ def from_objs(entities: Union[np.ndarray, list]) -> DataSubjectList:
...
+        # We need to project these data subject arrays to their entire row/column respectively
+        dsl1_projection = np.ones(
+            (*summed_dsl1.shape[:-1], summed_dsl2.shape[-1])
+        )  # *summed_dsl2.shape[1:-2],
+        dsl2_projection = np.ones(
+            (*summed_dsl2.shape[:-2], summed_dsl1.shape[-2], summed_dsl2.shape[-1])
+        )  # *summed_dsl2.shape[1:-2],
+
+        summed_dsl1.data_subjects_indexed = (
+            dsl1_projection * summed_dsl1.data_subjects_indexed
```
[the remark this answers is not in the document]
Comment (rasswanth-s) [reply packages/syft/src/syft/core/adp/data_subject_list.py]: Same as above, this DataSubjectList is stale

File: packages/syft/src/syft/core/adp/data_subject_list.py [hunk 8]
```diff
@@ -115,30 +147,674 @@ def from_objs(entities: Union[np.ndarray, list]) -> DataSubjectList:
...
+
+        output_ds = DataSubjectList.combine_dsi(summed_dsl1, summed_dsl2)
+
+        # This gets rid of redundant (repeating) NewDataSubject slices.
+        output_ds.data_subjects_indexed = np.unique(
+            output_ds.data_subjects_indexed, axis=0
+        )
+        return output_ds
+
+    def dot(dsl1: DataSubjectList, dsl2: DataSubjectList):
```
[the remark this answers is not in the document]
Comment (rasswanth-s) [reply packages/syft/src/syft/core/adp/data_subject_list.py]: Same as above, this DataSubjectList is stale

File: packages/syft/src/syft/core/adp/data_subject_list.py [hunk 9]
```diff
@@ -115,30 +147,674 @@ def from_objs(entities: Union[np.ndarray, list]) -> DataSubjectList:
...
+        return output_ds
+
+    @staticmethod
+    def index_dsl(tensor: Any, index):
+        if tensor.shape == tensor.data_subjects.shape:
+            return tensor.data_subjects[index]
+        elif len(tensor.shape) < len(tensor.data_subjects.shape):
+            index = tuple([slice(None, None, None)] + list(index))
+            return tensor.data_subjects[index]
+
```
[the remark this answers is not in the document]
Comment (rasswanth-s) [reply packages/syft/src/syft/core/adp/data_subject_list.py]: Same as above, this DataSubjectList is stale

[the remark this answers is not in the document]
Comment (rasswanth-s) [reply packages/syft/src/syft/core/adp/data_subject_list.py]: Old Code, we have removed it.

File: packages/syft/src/syft/core/adp/data_subject_list.py [hunk 10]
```diff
@@ -115,30 +147,674 @@ def from_objs(entities: Union[np.ndarray, list]) -> DataSubjectList:
...
+            output_shape = (dsl1.shape[0] + dsl2.shape[0], *dsl1.shape[1:])
+        elif dsl2_dims - dsl1_dims == 1:
+            dsl1.data_subjects_indexed = np.expand_dims(dsl1.data_subjects_indexed, 0)
+            output_shape = (dsl1.shape[0] + dsl2.shape[0], *dsl1.shape[1:])
+        else:
+            raise Exception(
+                f"Shapes of NewDataSubjects incompatible- are they meant to be broadcasted: {dsl1.shape}, {dsl2.shape}"
+            )
+
+        output_dsl = DataSubjectList(
```
[the remark this answers is not in the document]
Comment (rasswanth-s) [reply packages/syft/src/syft/core/adp/data_subject_list.py]: Same as above, this DataSubjectList is stale

File: packages/syft/src/syft/core/adp/data_subject_list.py [hunk 11]
```diff
@@ -115,30 +147,674 @@ def from_objs(entities: Union[np.ndarray, list]) -> DataSubjectList:
...
+            one_hot_lookup=np.array([]), data_subjects_indexed=np.empty(output_shape)
+        )
+
+        output_dsl[: dsl1.shape[0]] = dsl1
+        output_dsl[dsl1.shape[0] :] = dsl2
+        output_dsl.data_subjects_indexed = np.unique(
```
[the remark this answers is not in the document]
Comment (rasswanth-s) [reply packages/syft/src/syft/core/adp/data_subject_list.py]: Same as above, this DataSubjectList is stale.

File: packages/syft/src/syft/core/adp/data_subject_list.py [hunk 12]
```diff
@@ -115,30 +147,674 @@ def from_objs(entities: Union[np.ndarray, list]) -> DataSubjectList:
...
+            one_hot_lookup=self.one_hot_lookup,  # np.unique(self.one_hot_lookup[result]),
+            data_subjects_indexed=result,
+        )
+
+    def __setitem__(self, key, value) -> None:
+        if isinstance(value, DataSubjectList):
+            overlapping_ds = np.isin(value.one_hot_lookup, self.one_hot_lookup)
+            copied = value.copy()
+            if any(overlapping_ds):
+                # Find which Data Subjects have been assigned an index already
```
[the remark this answers is not in the document]
Comment (rasswanth-s) [reply packages/syft/src/syft/core/adp/data_subject_list.py]: Same as above, this DataSubjectList is stale

**Review by iamtrask — APPROVED** [review]
Well done all. This was a monster effort and adds some serious functionality. Congratulations and great work!

**Comment by madhavajay** [comment]
Outstanding work. @IshanMi @shubham3121 @rasswanth-s @IonesioJunior  This PR contains some significantly impressive work combining DL + DP that likely doesn't exist anywhere else. Im sure Copilot will be learning to reproduce this in no time. 😂
